pkg/cloudprovider/vsphere/loadbalancer/config/config.go

```diff
@@ -0,0 +1,297 @@
+	AdditionalTags      map[string]string                   `gcfg:"Tags"`
+	NSXTSimulation      *NsxtSimulation                     `gcfg:"NSX-T-Simulation"`
+}
+
+// LoadBalancerConfig contains the configuration for the load balancer itself
+type LoadBalancerConfig struct {
+	IPPoolName      string `gcfg:"ipPoolName"`
+	IPPoolID        string `gcfg:"ipPoolID"`
+	Size            string `gcfg:"size"`
+	LBServiceID     string `gcfg:"lbServiceId"`
+	LogicalRouterID string `gcfg:"logicalRouterId"`
+}
+
+// LoadBalancerClassConfig contains the configuration for a load balancer class
+type LoadBalancerClassConfig struct {
+	IPPoolName string `gcfg:"ipPoolName"`
+	IPPoolID   string `gcfg:"ipPoolID"`
+}
+
+// NsxtConfig contains the NSX-T specific configuration
+type NsxtConfig struct {
+	// NSX-T username.
+	User string `gcfg:"user"`
+	// NSX-T password in clear text.
+	Password string `gcfg:"password"`
+	// NSX-T host.
+	Host string `gcfg:"host"`
+	// True if vCenter uses self-signed cert.
+	InsecureFlag       bool   `gcfg:"insecure-flag"`
+	RemoteAuth         bool   `gcfg:"remote-auth"`
+	MaxRetries         int    `gcfg:"max-retries"`
+	RetryMinDelay      int    `gcfg:"retry-min-delay"`
+	RetryMaxDelay      int    `gcfg:"retry-max-delay"`
+	RetryOnStatusCodes []int  `gcfg:"retry-on-status-codes"`
+	ClientAuthCertFile string `gcfg:"client-auth-cert-file"`
+	ClientAuthKeyFile  string `gcfg:"client-auth-key-file"`
+	CAFile             string `gcfg:"ca-file"`
+}
+
+// NsxtSimulation is a helper configuration to pass fake data for testing purposes
+type NsxtSimulation struct {
+	SimulatedIPPools []string `gcfg:"simulatedIPPools"`
+}
+
+// IsEnabled checks whether the load balancer feature is enabled
+func (cfg *LBConfig) IsEnabled() bool {
+	return len(cfg.LoadBalancerClasses) > 0 || !cfg.LoadBalancer.IsEmpty()
+}
+
+func (cfg *LBConfig) validateConfig() error {
+	if cfg.LoadBalancer.LBServiceID == "" && cfg.LoadBalancer.LogicalRouterID == "" {
+		msg := "load balancer servive id or logical router id required"
+		klog.Errorf(msg)
+		return fmt.Errorf(msg)
+	}
+	if _, ok := SizeToMaxVirtualServers[cfg.LoadBalancer.Size]; !ok {
+		msg := "load balancer size is invalid"
+		klog.Errorf(msg)
+		return fmt.Errorf(msg)
+	}
+	if cfg.LoadBalancer.IPPoolID == "" && cfg.LoadBalancer.IPPoolName == "" {
+		class, ok := cfg.LoadBalancerClasses[DefaultLoadBalancerClass]
+		if !ok {
+			msg := "no default load balancer class defined"
+			klog.Errorf(msg)
+			return fmt.Errorf(msg)
+		} else if class.IPPoolName == "" && class.IPPoolID == "" {
+			msg := "default load balancer class: ipPoolName and ipPoolID is empty"
+			klog.Errorf(msg)
+			return fmt.Errorf(msg)
+		}
+	} else {
+		if cfg.LoadBalancer.IPPoolName == "" && cfg.LoadBalancer.IPPoolID == "" {
+			msg := "load balancer ipPoolName and ipPoolID is empty"
+			klog.Errorf(msg)
+			return fmt.Errorf(msg)
+		}
+	}
+	if cfg.NSXTSimulation == nil {
+		return cfg.NSXT.validateConfig()
```

Comment:
Hmm, the idea was to be able to call the validation function without a prior call of the `FromEnv` function. Because all this is done only once during startup it shouldn't be a critical issue.

Reply:
rearranged